garrysmod/lua/vgui/dcombobox.lua

```diff
@@ -1,289 +1,355 @@
-			end
-		end
-	end
-
-	local x, y = self:LocalToScreen( 0, self:GetTall() )
-
-	self.Menu:SetMinimumWidth( self:GetWide() )
-	self.Menu:Open( x, y, false, self )
-
-	self:OnMenuOpened( self.Menu )
-
-end
-
-function PANEL:CloseMenu()
-
-	if ( IsValid( self.Menu ) ) then
-		self.Menu:Remove()
-	end
-
-end
-
--- This really should use a convar change hook
-function PANEL:CheckConVarChanges()
-
-	if ( !self.m_strConVar ) then return end
-
-	local strValue = GetConVarString( self.m_strConVar )
-	if ( self.m_strConVarValue == strValue ) then return end
-
-	self.m_strConVarValue = strValue
-
-	self:SetValue( self:GetOptionTextByData( self.m_strConVarValue ) )
-
-end
-
-function PANEL:Think()
-
-	self:CheckConVarChanges()
-
-end
-
-function PANEL:SetValue( strValue )
-
-	self:SetText( strValue )
-
-end
-
-function PANEL:DoClick()
-
-	if ( self:IsMenuOpen() ) then
-		return self:CloseMenu()
-	end
-
-	self:OpenMenu()
-
-end
-
-function PANEL:GenerateExample( ClassName, PropertySheet, Width, Height )
-
-	local ctrl = vgui.Create( ClassName )
-	ctrl:AddChoice( "Some Choice" )
-	ctrl:AddChoice( "Another Choice", "myData" )
-	ctrl:AddChoice( "Default Choice", "myData2", true )
-	ctrl:AddChoice( "Icon Choice", "myData3", false, "icon16/star.png" )
-	ctrl:SetWide( 150 )
-
-	PropertySheet:AddSheet( ClassName, ctrl, nil, true, true )
-
-end
-
-derma.DefineControl( "DComboBox", "", PANEL, "DButton" )
+
+local PANEL = {}
+
+Derma_Hook( PANEL, "Paint", "Paint", "ComboBox" )
+
+Derma_Install_Convar_Functions( PANEL )
+
+AccessorFunc( PANEL, "m_bDoSort", "SortItems", FORCE_BOOL )
+
+function PANEL:Init()
+
+	-- Create data tables
+	self.Choices = {}
+	self.Data = {}
+	self.ChoiceIcons = {}
+	self.Spacers = {}
+
+	-- Create button
+	self.DropButton = vgui.Create( "DPanel", self )
+	self.DropButton.Paint = function( panel, w, h ) derma.SkinHook( "Paint", "ComboDownArrow", panel, w, h ) end
+	self.DropButton:SetMouseInputEnabled( false )
+	self.DropButton.ComboBox = self
+
+	-- Apply default panel hight
+	self:SetTall( 22 )
+
+	-- Setup internals
+	self:SetContentAlignment( 4 )
+	self:SetTextInset( 8, 0 )
+	self:SetIsMenu( true )
+	self:SetSortItems( true )
+
+end
+
+function PANEL:CloseMenu()
+
+	if ( IsValid( self.Menu ) ) then
+		self.Menu:Remove()
+	end
+
+	self.Menu = nil
+
+end
+
+function PANEL:Clear()
+
+	-- Does not allocate another table
+	self:SetText( "" )
+	table.Empty( self.Choices )
+	table.Empty( self.Data )
+	table.Empty( self.ChoiceIcons )
+	table.Empty( self.Spacers )
+	self.selected = nil
+	self:CloseMenu()
+
+end
+
+function PANEL:GetSelectedID()
+
+	return self.selected
+
+end
+
+function PANEL:GetChoiceCount()
+
+	return #self.Choices
+
+end
+
+function PANEL:GetOptionText( id )
+
+	return self.Choices[ id ]
+
+end
+
+function PANEL:GetOptionData( id )
+
+	return self.Data[ id ]
+
+end
+
+function PANEL:GetOptionIcon( id )
+
+	return self.ChoiceIcons[ id ]
+
+end
+
+function PANEL:GetOptionSpacer( id )
+
+	return self.Spacers[ id ]
+
+end

```

Comment:
These functions serve no purpose. This is a GUI element, not a data structure. There's no reason to ever want to get this data.